[Clang] Fix 'counted_by' for nested struct pointers

Cydox opened the pull request:
Fixes #110385

Fix counted_by attribute for cases where the flexible array member is accessed through struct pointer inside another struct:

```
struct variable {
        int a;
        int b;
        int length;
        short array[] __attribute__((counted_by(length)));
};

struct bucket {
        int a;
        struct variable *growable;
        int b;
};
```

__builtin_dynamic_object_size(p->growable->array, 0);

This commit makes sure that if the StructBase is both a MemberExpr and a pointer, it is treated as a pointer. Otherwise clang will generate to code to access the address of p->growable intead of loading the value of p->growable->length.

Cydox committed:
[Clang] Fix 'counted_by' for nested struct pointers

Fixes #110385

Fix counted_by attribute for cases where the flexible array member is accessed
through struct pointer inside another struct:

struct variable {
        int a;
        int b;
        int length;
        short array[] __attribute__((counted_by(length)));
};

struct bucket {
        int a;
        struct variable *growable;
        int b;
};

__builtin_dynamic_object_size(p->growable->array, 0);

This commit makes sure that if the StructBase is both a MemberExpr and a
pointer, it is treated as a pointer. Otherwise clang will generate to
code to access the address of p->growable intead of loading the value of
p->growable->length.

efriedma-quic commented:
(See discussion on #110487)

Cydox committed:
[Clang] cleanup CodeGenFunction::EmitLoadOfCountedByField

The DeclRefExpr case can also be handled by the pointer case

Cydox commented:
Removed the special case for DeclRefExpr, as this can also be handled by the pointer case. A few were adjusted, but none of them changed any behavior.

That additional commit is not necessary to fix anything, it's just cleanup.

efriedma-quic reviewed (commented):
StructAccessBase has two possible results: it finds an lvalue that's the base, which can't be peeled apart any further, or it finds a pointer that's the base.  It shouldn't be possible to confuse a pointer with a MemberExpr: a MemberExpr is an lvalue, and a pointer is, in this context, an rvalue.

So I think the issue here is that StructAccessBase is returning the wrong result, and trying to paper over that is going to lead to issues that are harder to track down later.

Cydox commented:
> StructAccessBase has two possible results: it finds an lvalue that's the base, which can't be peeled apart any further, or it finds a pointer that's the base. It shouldn't be possible to confuse a pointer with a MemberExpr: a MemberExpr is an lvalue, and a pointer is, in this context, an rvalue.

To confirm:
Are you saying that StructBase can't be both a MemberExpr and a pointer?

For context: I'm not familiar with the clang/llvm internals before I looked into this issue

Cydox commented:
Alright I think I'm getting what you mean with skipping over the LValueToRValue cast. I'll try to put a better fix together...

bwendling commented:
The problem we're faced with here is that the `Base` pointer could point to anywhere within the structure. We already jump through several hoops to get the flexible array member's `Decl` and the counter's `Decl`.

So because `Base` could be a pointer to anywhere in the struct, we have to determine *where* it's pointing. If it's a pointer to the beginning of the struct (i.e. `__builtin_dynamic_object_size(ptr, 0)`), then we expect a `DeclRefExpr` from the visitor. But we could have something like this:

```c
struct s {
  struct s *p;
  /* ... */
} *ptr;

__builtin_dynamic_object_size(ptr->p->p->p->p, 0);
```

That's caught as a `MemberExpr`. We can have the `__bdos` pointing somewhere within the flexible array member. Etc., etc. Notice that we can't rely upon looking at a `CastExpr's` type to determine exactly what we're looking at.

What this code is therefore *trying* to do is simply find a starting point (some `Expr`) from which to build a GEP to the `counted_by` field. It does that in part by using the `CountDecl's` outer lexical record context to stop when we've found the correct base (note that the counted_by's outer lexical record context has already been checked to be the same as the flexible array member's outer lexical record context).

I'm not sure what you mean by "or it finds a pointer that's the base". The `MemberExpr` the visitor returns doesn't have to be a pointer---e.g. `ptr->a.b.c.d` returning `ptr->a.b`, where `b` isn't a pointer---or do you mean something else?

efriedma-quic commented:
> I'm not sure what you mean by "or it finds a pointer that's the base". The MemberExpr the visitor returns doesn't have to be a pointer---e.g. ptr->a.b.c.d returning ptr->a.b, where b isn't a pointer---or do you mean something else?

In an expression which does not contain "->" operators, StructAccessBase should return an lvalue: the expression which describes the base of the relevant struct.  There might be addressof/dereference/array decay/etc. in the middle, but that's the ultimate result.  The "or it finds a pointer" bit isn't relevant in this case.

"LHS->member" is equivalent to "(*LHS).member"... and sometimes StructAccessBase wants to return "*LHS".  This is where the "or it finds a pointer" bit comes in; there is no expression *LHS in the AST, so it returns a pointer which needs to be dereferenced to turn it into an lvalue.

Cydox commented:
I'm gonna push something real quick to see if I understand were you want us to go. Didn't check if this breaks any tests yet.

Cydox committed:
no special case for MemberExpr, just LValue

Cydox commented:
@efriedma-quic do you mean somthing along those lines?

bwendling commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -10,6 +10,7 @@
 //
 //===----------------------------------------------------------------------===//
 
+#include <iostream>
```
Please don't include this (I know it's left from debugging). If you want to debug things, you can use:

```c++
llvm::errs() << "debug message\n";
```

Cydox replied:
Yeah, this is just a quick commit to show you where my mind is going right now. Won't be in any final commit

Cydox commented:
So, this doesn't fully work yet. When compiling the kernel it seems like there are cases were `StructBase` is now both an LValue and a pointer. Still investigating.

Cydox commented:
Yeah so the problem is if you do `__builtin_dynamic_object_size(v, 0)`

In that case it's a `DeclRefExpr`, a pointer, and an `LValue`.

Easy fix is to check if `StructBase` is a pointer instead of checking for LValue, but then I don't really see the point of this change vs my originial PR to be honest (maybe I'm missing something).

Cydox committed:
check for pointer, not LValue

Cydox commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -1164,15 +1163,15 @@ llvm::Value *CodeGenFunction::EmitLoadOfCountedByField(
     TBAAAccessInfo TBAAInfo;
     Address Addr = EmitPointerWithAlignment(StructBase, &BaseInfo, &TBAAInfo);
     Res = Addr.emitRawPointer(*this);
+  } else {
```
should I add a check for `StructBase->isLValue()` and add an `else` statement that returns `nullptr`?

Cydox replied:
added that extra check

Cydox commented:
Now this passes all tests again, also compiles the kernel without issue.

Cydox committed:
check if StructBase is LValue

bwendling reviewed (approved):
Thanks for the patch!

efriedma-quic commented:
> Yeah so the problem is if you do `__builtin_dynamic_object_size(v, 0)`
> 
> In that case it's a `DeclRefExpr`, a pointer, and an `LValue`.

Can you give a more complete example?  I just tried the following, and I see an lvaluetorvalue cast.

int f(const void *p) { return __builtin_dynamic_object_size(p, 0); }

Cydox commented:
> > Yeah so the problem is if you do `__builtin_dynamic_object_size(v, 0)`
> > In that case it's a `DeclRefExpr`, a pointer, and an `LValue`.
> 
> Can you give a more complete example? I just tried the following, and I see an lvaluetorvalue cast.
> 
> ```
> int f(const void *p) { return __builtin_dynamic_object_size(p, 0); }
> ```

```C
// test2.c

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

struct variable {
        int a;
        int b;
        int length;
        short array[] __attribute__((counted_by(length)));
};

int main(int argc, char *argv[])
{
        struct variable *v;

        v = malloc(sizeof(struct variable) + sizeof(short) * 32);
        v->length = 32;


        printf("%zu\n", __builtin_dynamic_object_size(v, 0));

        return 0;
}
```

I added this `StructBase->dump`:
```
diff --git a/clang/lib/CodeGen/CGExpr.cpp b/clang/lib/CodeGen/CGExpr.cpp
index 4aca60685f37..7a06819f1a67 100644
--- a/clang/lib/CodeGen/CGExpr.cpp
+++ b/clang/lib/CodeGen/CGExpr.cpp
@@ -1162,6 +1162,8 @@ llvm::Value *CodeGenFunction::EmitLoadOfCountedByField(
   if (!StructBase || StructBase->HasSideEffects(getContext()))
     return nullptr;
 
+  StructBase->dump();
+
   llvm::Value *Res = nullptr;
   if (StructBase->getType()->isPointerType()) {
     LValueBaseInfo BaseInfo;

```

```
$ clang test2.c
DeclRefExpr 0x34b302c8 'struct variable *' lvalue Var 0x34b2fce8 'v' 'struct variable *'

$ ./a.out
76
```

In my testing I also added a print statement that printed a dump when both `StructBase->getType()->isPointerType()` and `StructBase->isLValue()` were true and there were a bunch of hits when compiling the kernel.

efriedma-quic commented:
ast-dump for your example:

```
|   | `-CallExpr 0xcc44978 <col:25, col:59> 'unsigned long'
|   |   |-ImplicitCastExpr 0xcc44960 <col:25> 'unsigned long (*)(const void *, int)' <BuiltinFnToFnPtr>
|   |   | `-DeclRefExpr 0xcc448b8 <col:25> '<builtin fn type>' Function 0xcc446b0 '__builtin_dynamic_object_size' 'unsigned long (const void *, int)'
|   |   |-ImplicitCastExpr 0xcc449c0 <col:55> 'const void *' <BitCast>
|   |   | `-ImplicitCastExpr 0xcc449a8 <col:55> 'struct variable *' <LValueToRValue>
|   |   |   `-DeclRefExpr 0xcc448d8 <col:55> 'struct variable *' lvalue Var 0xcc41278 'v' 'struct variable *'
|   |   `-IntegerLiteral 0xcc448f8 <col:58> 'int' 0
```

I guess CodeGenFunction::emitFlexibleArrayMemberSize calls IgnoreParenImpCasts() which throws away the lvalue-to-rvalue.

efriedma-quic commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -1095,6 +1095,8 @@ class StructAccessBase
   }
   const Expr *VisitCastExpr(const CastExpr *E) {
+    if (E->getCastKind() == CK_LValueToRValue)
+      return E;
```
Should this be `return IsExpectedRecordDecl(E) ? E : nullptr;`?

bwendling replied:
It might be a good sanity check, though I think I've only seen an `LValueToRValue` cast before the node we wanted.

bwendling replied:
See https://github.com/llvm/llvm-project/pull/111448
